# Titanic Survival

In [46]:
import os
os.getcwd()

'C:\\Users\\kbala\\Assignment_3\\src'

In [47]:
from IPython.core.display import HTML
HTML(filename= "../data/titanic.html")

In [48]:
import pandas as pd

# you would need xlrd - pip install xlrd
t_file = pd.ExcelFile("../data/titanic.xls")
t_df = t_file.parse("titanic", header=0)
t_df.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1   0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [49]:
t_df.size

18326

In [50]:
t_df.shape

(1309, 14)

In [51]:
survived_titanic = t_df.loc[t_df['survived'] == 1]

In [52]:
survived_titanic

pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
1          1         1                   Allison, Master. Hudson Trevor   
5          1         1                              Anderson, Mr. Harry   
6          1         1                Andrews, Miss. Kornelia Theodosia   
8          1         1    Appleton, Mrs. Edward Dale (Charlotte Lamson)   
...      ...       ...                                              ...   
1261       3         1                           Turkula, Mrs. (Hedwig)   
1277       3         1                             Vartanian, Mr. David   
1286       3         1  Whabee, Mrs. George Joseph (Shawneene Abi-Saab)   
1290       3         1                 Wilkes, Mrs. James (Ellen Needs)   
1300       3         1          Yasbeck, Mrs. Antoni (Selini Alexander)   

         sex      age  sibsp  parch  ticket      fare    cabin embarked  \
0     female  29.0000      0      0   24160  211.3375       B5        S   
1       male   0.9167      1      2  113781  151.5500  C22 C26        S   
5       male  48.0000      0      0   19952   26.5500      E12        S   
6     female  63.0000      1      0   13502   77.9583       D7        S   
8     female  53.0000      2      0   11769   51.4792     C101        S   
...      ...      ...    ...    ...     ...       ...      ...      ...   
1261  female  63.0000      0      0    4134    9.5875      NaN        S   
1277    male  22.0000      0      0    2658    7.2250      NaN        C   
1286  female  38.0000      0      0    2688    7.2292      NaN        C   
1290  female  47.0000      1      0  363272    7.0000      NaN        S   
1300  female  15.0000      1      0    2659   14.4542      NaN        C   

       boat  body                        home.dest  
0         2   NaN                     St Louis, MO  
1        11   NaN  Montreal, PQ / Chesterville, ON  
5         3   NaN                     New York, NY  
6        10   NaN                       Hudson, NY  
8         D   NaN              Bayside, Queens, NY  
...     ...   ...                              ...  
1261     15   NaN                              NaN  
1277  13 15   NaN                              NaN  
1286      C   NaN                              NaN  
1290    NaN   NaN                              NaN  
1300    NaN   NaN                              NaN  

[500 rows x 14 columns]

# Task 1:

## Use the groupby method to calculate the proportion of passengers that survived by sex

In [53]:
t_df.groupby(["sex"])["survived"].count()

sex
female    466
male      843
Name: survived, dtype: int64

In [54]:
survived_titanic.groupby(["sex"])["survived"].count()

sex
female    339
male      161
Name: survived, dtype: int64

In [55]:
(survived_titanic.groupby(["sex"])["survived"].count())/(t_df.groupby(["sex"])["survived"].count())

sex
female    0.727468
male      0.190985
Name: survived, dtype: float64

### Result : 
From above we can say 72% of females has survived whereas for men's it is 19%

# Task 2:

## Calculate the same proportion, but by class and sex. 

In [56]:
survived_titanic.groupby(["sex","pclass"])["survived"].count()

sex     pclass
female  1         139
        2          94
        3         106
male    1          61
        2          25
        3          75
Name: survived, dtype: int64

In [57]:
t_df.groupby(["sex","pclass"])["survived"].count()

sex     pclass
female  1         144
        2         106
        3         216
male    1         179
        2         171
        3         493
Name: survived, dtype: int64

In [58]:
(survived_titanic.groupby(["sex","pclass"])["survived"].count())/(t_df.groupby(["sex","pclass"])["survived"].count())

sex     pclass
female  1         0.965278
        2         0.886792
        3         0.490741
male    1         0.340782
        2         0.146199
        3         0.152130
Name: survived, dtype: float64

### Result:
From the above proportion we can say that if a female is from 1st class, they can high probablity of surviving and male from 2nd class has the least.


| sex      | pclass |proportion
| ----------- | ----------- |-----------|
|female  |1         |0.965278
|  |2         |0.886792
|  |3         |0.490741
|male  |1         |0.340782
|  |2         |0.146199
|  |3         |0.152130

# Task 3:

## Create age categories: children (under 14 years), adolescents (14-20), adult (21-64), and senior(65+), and calculate survival proportions by age category, class and sex. 

In [59]:
t_df.loc[t_df["age"] <= 14, 'age category'] = 'children'

t_df

pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
1          1         1                   Allison, Master. Hudson Trevor   
2          1         0                     Allison, Miss. Helen Loraine   
3          1         0             Allison, Mr. Hudson Joshua Creighton   
4          1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1304       3         0                             Zabour, Miss. Hileni   
1305       3         0                            Zabour, Miss. Thamine   
1306       3         0                        Zakarian, Mr. Mapriededer   
1307       3         0                              Zakarian, Mr. Ortin   
1308       3         0                               Zimmerman, Mr. Leo   

         sex      age  sibsp  parch  ticket      fare    cabin embarked boat  \
0     female  29.0000      0      0   24160  211.3375       B5        S    2   
1       male   0.9167      1      2  113781  151.5500  C22 C26        S   11   
2     female   2.0000      1      2  113781  151.5500  C22 C26        S  NaN   
3       male  30.0000      1      2  113781  151.5500  C22 C26        S  NaN   
4     female  25.0000      1      2  113781  151.5500  C22 C26        S  NaN   
...      ...      ...    ...    ...     ...       ...      ...      ...  ...   
1304  female  14.5000      1      0    2665   14.4542      NaN        C  NaN   
1305  female      NaN      1      0    2665   14.4542      NaN        C  NaN   
1306    male  26.5000      0      0    2656    7.2250      NaN        C  NaN   
1307    male  27.0000      0      0    2670    7.2250      NaN        C  NaN   
1308    male  29.0000      0      0  315082    7.8750      NaN        S  NaN   

       body                        home.dest age category  
0       NaN                     St Louis, MO          NaN  
1       NaN  Montreal, PQ / Chesterville, ON     children  
2       NaN  Montreal, PQ / Chesterville, ON     children  
3     135.0  Montreal, PQ / Chesterville, ON          NaN  
4       NaN  Montreal, PQ / Chesterville, ON          NaN  
...     ...                              ...          ...  
1304  328.0                              NaN          NaN  
1305    NaN                              NaN          NaN  
1306  304.0                              NaN          NaN  
1307    NaN                              NaN          NaN  
1308    NaN                              NaN          NaN  

[1309 rows x 15 columns]

Write condition where age is less than 14 make it as children.

Making remaining feilds similar to children based on age, adolescents (14-20), adult (21-64), and senior(65+),

In [60]:
t_df.loc[(t_df["age"].between(21, 64, inclusive = True)), 'age category'] = 'adults'
 
t_df

C:\Users\kbala\AppData\Local\Temp\ipykernel_16324\1792916240.py:1: FutureWarning: Boolean inputs to the `inclusive` argument are deprecated in favour of `both` or `neither`.
  t_df.loc[(t_df["age"].between(21, 64, inclusive = True)), 'age category'] = 'adults'


pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
1          1         1                   Allison, Master. Hudson Trevor   
2          1         0                     Allison, Miss. Helen Loraine   
3          1         0             Allison, Mr. Hudson Joshua Creighton   
4          1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1304       3         0                             Zabour, Miss. Hileni   
1305       3         0                            Zabour, Miss. Thamine   
1306       3         0                        Zakarian, Mr. Mapriededer   
1307       3         0                              Zakarian, Mr. Ortin   
1308       3         0                               Zimmerman, Mr. Leo   

         sex      age  sibsp  parch  ticket      fare    cabin embarked boat  \
0     female  29.0000      0      0   24160  211.3375       B5        S    2   
1       male   0.9167      1      2  113781  151.5500  C22 C26        S   11   
2     female   2.0000      1      2  113781  151.5500  C22 C26        S  NaN   
3       male  30.0000      1      2  113781  151.5500  C22 C26        S  NaN   
4     female  25.0000      1      2  113781  151.5500  C22 C26        S  NaN   
...      ...      ...    ...    ...     ...       ...      ...      ...  ...   
1304  female  14.5000      1      0    2665   14.4542      NaN        C  NaN   
1305  female      NaN      1      0    2665   14.4542      NaN        C  NaN   
1306    male  26.5000      0      0    2656    7.2250      NaN        C  NaN   
1307    male  27.0000      0      0    2670    7.2250      NaN        C  NaN   
1308    male  29.0000      0      0  315082    7.8750      NaN        S  NaN   

       body                        home.dest age category  
0       NaN                     St Louis, MO       adults  
1       NaN  Montreal, PQ / Chesterville, ON     children  
2       NaN  Montreal, PQ / Chesterville, ON     children  
3     135.0  Montreal, PQ / Chesterville, ON       adults  
4       NaN  Montreal, PQ / Chesterville, ON       adults  
...     ...                              ...          ...  
1304  328.0                              NaN          NaN  
1305    NaN                              NaN          NaN  
1306  304.0                              NaN       adults  
1307    NaN                              NaN       adults  
1308    NaN                              NaN       adults  

[1309 rows x 15 columns]

In [61]:
t_df.loc[(t_df["age"].between(14, 21, inclusive = True)), 'age category'] = 'adolescents'
 
t_df

C:\Users\kbala\AppData\Local\Temp\ipykernel_16324\1767548964.py:1: FutureWarning: Boolean inputs to the `inclusive` argument are deprecated in favour of `both` or `neither`.
  t_df.loc[(t_df["age"].between(14, 21, inclusive = True)), 'age category'] = 'adolescents'


pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
1          1         1                   Allison, Master. Hudson Trevor   
2          1         0                     Allison, Miss. Helen Loraine   
3          1         0             Allison, Mr. Hudson Joshua Creighton   
4          1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1304       3         0                             Zabour, Miss. Hileni   
1305       3         0                            Zabour, Miss. Thamine   
1306       3         0                        Zakarian, Mr. Mapriededer   
1307       3         0                              Zakarian, Mr. Ortin   
1308       3         0                               Zimmerman, Mr. Leo   

         sex      age  sibsp  parch  ticket      fare    cabin embarked boat  \
0     female  29.0000      0      0   24160  211.3375       B5        S    2   
1       male   0.9167      1      2  113781  151.5500  C22 C26        S   11   
2     female   2.0000      1      2  113781  151.5500  C22 C26        S  NaN   
3       male  30.0000      1      2  113781  151.5500  C22 C26        S  NaN   
4     female  25.0000      1      2  113781  151.5500  C22 C26        S  NaN   
...      ...      ...    ...    ...     ...       ...      ...      ...  ...   
1304  female  14.5000      1      0    2665   14.4542      NaN        C  NaN   
1305  female      NaN      1      0    2665   14.4542      NaN        C  NaN   
1306    male  26.5000      0      0    2656    7.2250      NaN        C  NaN   
1307    male  27.0000      0      0    2670    7.2250      NaN        C  NaN   
1308    male  29.0000      0      0  315082    7.8750      NaN        S  NaN   

       body                        home.dest age category  
0       NaN                     St Louis, MO       adults  
1       NaN  Montreal, PQ / Chesterville, ON     children  
2       NaN  Montreal, PQ / Chesterville, ON     children  
3     135.0  Montreal, PQ / Chesterville, ON       adults  
4       NaN  Montreal, PQ / Chesterville, ON       adults  
...     ...                              ...          ...  
1304  328.0                              NaN  adolescents  
1305    NaN                              NaN          NaN  
1306  304.0                              NaN       adults  
1307    NaN                              NaN       adults  
1308    NaN                              NaN       adults  

[1309 rows x 15 columns]

In [62]:
t_df.loc[t_df["age"] > 65, 'age category'] = 'senior'

t_df

pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
1          1         1                   Allison, Master. Hudson Trevor   
2          1         0                     Allison, Miss. Helen Loraine   
3          1         0             Allison, Mr. Hudson Joshua Creighton   
4          1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1304       3         0                             Zabour, Miss. Hileni   
1305       3         0                            Zabour, Miss. Thamine   
1306       3         0                        Zakarian, Mr. Mapriededer   
1307       3         0                              Zakarian, Mr. Ortin   
1308       3         0                               Zimmerman, Mr. Leo   

         sex      age  sibsp  parch  ticket      fare    cabin embarked boat  \
0     female  29.0000      0      0   24160  211.3375       B5        S    2   
1       male   0.9167      1      2  113781  151.5500  C22 C26        S   11   
2     female   2.0000      1      2  113781  151.5500  C22 C26        S  NaN   
3       male  30.0000      1      2  113781  151.5500  C22 C26        S  NaN   
4     female  25.0000      1      2  113781  151.5500  C22 C26        S  NaN   
...      ...      ...    ...    ...     ...       ...      ...      ...  ...   
1304  female  14.5000      1      0    2665   14.4542      NaN        C  NaN   
1305  female      NaN      1      0    2665   14.4542      NaN        C  NaN   
1306    male  26.5000      0      0    2656    7.2250      NaN        C  NaN   
1307    male  27.0000      0      0    2670    7.2250      NaN        C  NaN   
1308    male  29.0000      0      0  315082    7.8750      NaN        S  NaN   

       body                        home.dest age category  
0       NaN                     St Louis, MO       adults  
1       NaN  Montreal, PQ / Chesterville, ON     children  
2       NaN  Montreal, PQ / Chesterville, ON     children  
3     135.0  Montreal, PQ / Chesterville, ON       adults  
4       NaN  Montreal, PQ / Chesterville, ON       adults  
...     ...                              ...          ...  
1304  328.0                              NaN  adolescents  
1305    NaN                              NaN          NaN  
1306  304.0                              NaN       adults  
1307    NaN                              NaN       adults  
1308    NaN                              NaN       adults  

[1309 rows x 15 columns]

As there are some NaN values, making those as Adults for calculation purpose

In [63]:
t_df.loc[t_df["age"] == 'NaN', 'age category'] = 'adults'

t_df

pclass  survived                                             name  \
0          1         1                    Allen, Miss. Elisabeth Walton   
1          1         1                   Allison, Master. Hudson Trevor   
2          1         0                     Allison, Miss. Helen Loraine   
3          1         0             Allison, Mr. Hudson Joshua Creighton   
4          1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
...      ...       ...                                              ...   
1304       3         0                             Zabour, Miss. Hileni   
1305       3         0                            Zabour, Miss. Thamine   
1306       3         0                        Zakarian, Mr. Mapriededer   
1307       3         0                              Zakarian, Mr. Ortin   
1308       3         0                               Zimmerman, Mr. Leo   

         sex      age  sibsp  parch  ticket      fare    cabin embarked boat  \
0     female  29.0000      0      0   24160  211.3375       B5        S    2   
1       male   0.9167      1      2  113781  151.5500  C22 C26        S   11   
2     female   2.0000      1      2  113781  151.5500  C22 C26        S  NaN   
3       male  30.0000      1      2  113781  151.5500  C22 C26        S  NaN   
4     female  25.0000      1      2  113781  151.5500  C22 C26        S  NaN   
...      ...      ...    ...    ...     ...       ...      ...      ...  ...   
1304  female  14.5000      1      0    2665   14.4542      NaN        C  NaN   
1305  female      NaN      1      0    2665   14.4542      NaN        C  NaN   
1306    male  26.5000      0      0    2656    7.2250      NaN        C  NaN   
1307    male  27.0000      0      0    2670    7.2250      NaN        C  NaN   
1308    male  29.0000      0      0  315082    7.8750      NaN        S  NaN   

       body                        home.dest age category  
0       NaN                     St Louis, MO       adults  
1       NaN  Montreal, PQ / Chesterville, ON     children  
2       NaN  Montreal, PQ / Chesterville, ON     children  
3     135.0  Montreal, PQ / Chesterville, ON       adults  
4       NaN  Montreal, PQ / Chesterville, ON       adults  
...     ...                              ...          ...  
1304  328.0                              NaN  adolescents  
1305    NaN                              NaN          NaN  
1306  304.0                              NaN       adults  
1307    NaN                              NaN       adults  
1308    NaN                              NaN       adults  

[1309 rows x 15 columns]

In [64]:
t_df.groupby(["sex","pclass","age category"])["survived"].count()

sex     pclass  age category
female  1       adolescents      18
                adults          113
                children          1
                senior            1
        2       adolescents      15
                adults           74
                children         14
        3       adolescents      40
                adults           81
                children         31
male    1       adolescents       7
                adults          132
                children          5
                senior            5
        2       adolescents      23
                adults          122
                children         11
                senior            2
        3       adolescents      88
                adults          221
                children         37
                senior            2
Name: survived, dtype: int64

## Make the age category for survived_titanic table

In [65]:
survived_titanic = t_df.loc[t_df['survived'] == 1]

In [66]:
survived_titanic.groupby(["sex","pclass","age category"])["survived"].count()

sex     pclass  age category
female  1       adolescents      18
                adults          109
                senior            1
        2       adolescents      14
                adults           64
                children         14
        3       adolescents      20
                adults           37
                children         15
male    1       adolescents       2
                adults           45
                children          5
                senior            1
        2       adolescents       2
                adults           10
                children         11
        3       adolescents      12
                adults           35
                children         12
Name: survived, dtype: int64

In [67]:
(survived_titanic.groupby(["sex","pclass","age category"])["survived"].count())/(t_df.groupby(["sex","pclass","age category"])["survived"].count())

sex     pclass  age category
female  1       adolescents     1.000000
                adults          0.964602
                children             NaN
                senior          1.000000
        2       adolescents     0.933333
                adults          0.864865
                children        1.000000
        3       adolescents     0.500000
                adults          0.456790
                children        0.483871
male    1       adolescents     0.285714
                adults          0.340909
                children        1.000000
                senior          0.200000
        2       adolescents     0.086957
                adults          0.081967
                children        1.000000
                senior               NaN
        3       adolescents     0.136364
                adults          0.158371
                children        0.324324
                senior               NaN
Name: survived, dtype: float64

NaN values can be made as 0, which means none of those category people has survived.